lib/src/disposable/transformer/impl_generation.dart

```diff
@@ -0,0 +1,164 @@
+  if (onDisposeHandler == null) {
+    // Generate an `onDispose()` method.
+    endOfClassBuffer
+      ..write('@override Future<Null> onDispose() { ')
+      ..write('await super.onDispose(); ')
+      ..write('$nullOutFieldsMethodName(); ')
+      ..write('} ');
+  } else {
+    // Determine where to augment the existing `onDispose()` based on the type
+    // of function body.
+    if (onDisposeHandler.body is BlockFunctionBody) {
+      // For a block function, the call to null out fields will be inserted
+      // immediately before the right bracket on the same line.
+      BlockFunctionBody body = onDisposeHandler.body;
+      transformedFile.insert(
+        sourceFile.location(body.block.rightBracket.offset),
+        '$nullOutFieldsMethodName(); ',
+      );
+    } else if (onDisposeHandler.body is EmptyFunctionBody) {
+      // For an empty function, the body needs to be replaced with a block body
+      // implementation, but on one line.
+      EmptyFunctionBody body = onDisposeHandler.body;
+      transformedFile.replace(
+        sourceFile.span(
+          onDisposeHandler.parameters.rightParenthesis.end,
+          body.semicolon.end,
+        ),
+        ' async { await super.onDispose(); $nullOutFieldsMethodName(); }',
+      );
+    } else if (onDisposeHandler.body is ExpressionFunctionBody) {
+      // For an expression function (fat-arrow), the body needs to be changed
+      // to a block body implementation so the expression can be augmented.
+      ExpressionFunctionBody body = onDisposeHandler.body;
+
+      // Step 1: insert the `async` keyword if not present
+      if (body.keyword == null) {
+        transformedFile.insert(
+          sourceFile.location(body.functionDefinition.offset),
+          'async ',
+        );
+      }
+
+      // Step 2: replace the => with a {
+      transformedFile.replace(
+        sourceFile.span(
+          body.functionDefinition.offset,
+          body.functionDefinition.end,
+        ),
+        '{',
+      );
+
+      // Step 3: augment the expression. This requires inserting an `await`
+      // before the original expression to retain the original behavior (since
+      // whatever was being returned would have been awaited) and then inserting
+      // the call to null out the fields.
+      transformedFile.insert(
+        sourceFile.location(body.expression.offset),
+        'await ',
+      );
+      transformedFile.insert(
+        sourceFile.location(body.expression.end),
+        '; $nullOutFieldsMethodName(); ',
+      );
+
+      // Step 4: replace the ; with a }
+      transformedFile.replace(
+        sourceFile.span(
+          body.semicolon.offset,
+          body.semicolon.end,
+        ),
+        '}',
+      );
+    }
+  }
+
+  // Generate the method to null-out and clear all fields.
+  endOfClassBuffer.write('void $nullOutFieldsMethodName() { ');
+  for (var fieldName in clearableFieldNames) {
+    endOfClassBuffer.write('${fieldName}.clear(); ');
```

Comment:
Should this be null-aware? In theory the list / map might have been initialized to null.